org.eclipse.xtext.builder.standalone/src/org/eclipse/xtext/builder/standalone/incremental/ClasspathInfos.java

```diff
@@ -0,0 +1,74 @@
+/*******************************************************************************
+ * Copyright (c) 2022 Sebastian Zarnekow and others.
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *******************************************************************************/
+package org.eclipse.xtext.builder.standalone.incremental;
+
+import java.io.BufferedInputStream;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.nio.charset.StandardCharsets;
+import java.util.Arrays;
+import java.util.List;
+import java.util.concurrent.ConcurrentHashMap;
+
+import org.eclipse.core.runtime.IPath;
+
+import com.google.common.hash.Funnels;
+import com.google.common.hash.HashCode;
+import com.google.common.hash.Hasher;
+import com.google.common.io.ByteStreams;
+import com.google.common.io.Files;
+import com.google.inject.Singleton;
+
+/**
+ * @author Sebastian Zarnekow - Initial contribution and API
+ */
+@Singleton
+public class ClasspathInfos {
+	
+	private ConcurrentHashMap<IPath, HashCode> hashes = new ConcurrentHashMap<>();
+	
+	public byte[] hashClassesOrJar(IPath path) {
+		return hashes.computeIfAbsent(path, any->{
+			if ("jar".equalsIgnoreCase(path.getFileExtension())) {
+				List<String> segments = Arrays.asList(path.segments());
+				if (segments.contains(".gradle")) {
+					String maybeHash = segments.get(segments.size() - 2);
+					if (maybeHash.length() >= 36) {
+						return BinaryFileHashing.hashFunction().hashString(maybeHash, StandardCharsets.ISO_8859_1);
```

Comment:
hmmm. can we rely on iso?

Reply:
I think so, yes.
The gradle pathname is a sha1 hash which is represented with ascii chars that are valid in the path name:

https://docs.gradle.org/current/userguide/dependency_management.html and https://stackoverflow.com/questions/35571390/what-does-the-random-string-mean-as-sub-directory-names-in-gradle-cache-director